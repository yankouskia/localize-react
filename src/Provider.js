import React, { useEffect, useState } from 'react';
import {
  buildTranslation,
  clearCache,
  memoize,
  sanitizeLocale,
} from './helpers';

export const LocalizationContext = React.createContext();

export function LocalizationProvider({ children, disableCache, locale, translations = {} }) {

  const [currentLocale, setLocale] = useState(locale);

  useEffect(() => {
    if (currentLocale !== locale) {
      setLocale(locale)
    }
  }, [locale]);

  const sanitizedLocale = sanitizeLocale(currentLocale, translations);
  const pureTranslations = sanitizedLocale ? translations[sanitizedLocale] : translations;

  useEffect(() => {
    clearCache();
  }, [currentLocale, translations]);

  function pureTranslateFn(key, values, defaultMessage) {
    if (!pureTranslations || !key) return defaultMessage || key;

    const fallbackTranslation = typeof defaultMessage === 'string' ? defaultMessage : key;
    const possibleValue = pureTranslations[key];

    if (typeof possibleValue === 'string') {
      if (!values) return possibleValue;

      return buildTranslation(possibleValue, values);
    }

    const complexKeyArray = key.split('.');
    if (complexKeyArray.length === 1) return buildTranslation(fallbackTranslation, values);

    let finalValue = pureTranslations[complexKeyArray[0]];
    for (let i = 1; i < complexKeyArray.length; i++) {
      if (finalValue) {
        finalValue = finalValue[complexKeyArray[i]];
      }
    }

    return typeof finalValue === 'string' ? buildTranslation(finalValue, values) : buildTranslation(fallbackTranslation, values);
  }

  const translate = disableCache ? pureTranslateFn : memoize(pureTranslateFn);

  return (
    <LocalizationContext.Provider value={{ locale: currentLocale, translate, translations, setLocale }}>
      {children}
    </LocalizationContext.Provider>
  );
}

export const LocalizationConsumer = LocalizationContext.Consumer;
