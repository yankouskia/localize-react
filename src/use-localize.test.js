import React, { useEffect } from 'react';
import renderer from 'react-test-renderer';
import { LocalizationProvider } from './Provider';
import { useLocalize } from './use-localize';

const TRANSLATIONS = {
  name: 'Alex',
  en: {
    msg: 'English'
  },
  ja: {
    msg: 'Japanese'
  }
};

describe('Localize hook', () => {
  it('translates from key to text', () => {
    function Test() {
      const { translate } = useLocalize();

      return translate('name');
    }

    const tree = renderer
      .create(
        <LocalizationProvider
          translations={TRANSLATIONS}
        >
          <Test />
        </LocalizationProvider>
      )
      .toJSON();

      expect(tree).toMatchSnapshot();
  });

  it('changes locale', () => {
    function Test() {
      const { translate, setLocale } = useLocalize();

      useEffect(() => {
        setLocale("ja");
      }, []);

      return translate('msg');
    }

    let tree;

    renderer.act(() => {
      tree = renderer.create(
        <LocalizationProvider
          locale={"en"}
          translations={TRANSLATIONS}
        >
          <Test />
        </LocalizationProvider>
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
