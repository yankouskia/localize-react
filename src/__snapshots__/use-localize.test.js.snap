// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Localize hook changes locale 1`] = `"Japanese"`;

exports[`Localize hook translates from key to text 1`] = `"Alex"`;
